cf_xarray/geometry.py

```diff
@@ -293,3 +293,64 @@ def cf_to_points(ds: xr.Dataset):
         j += n
 
     return xr.DataArray(geoms, dims=node_count.dims, coords=node_count.coords)
+
+
+def grid_bounds_to_polygons(ds: xr.Dataset) -> xr.DataArray:
+    """
+    Converts a regular 2D lat/lon grid to a 2D array of shapely polygons.
+
+    Modified from https://notebooksharing.space/view/c6c1f3a7d0c260724115eaa2bf78f3738b275f7f633c1558639e7bbd75b31456.
+
+    Parameters
+    ----------
+    ds : xr.Dataset
+        Dataset with "latitude" and "longitude" variables as well as their bounds variables.
+        1D "latitude" and "longitude" variables are supported. This function will automatically
+        broadcast them against each other.
+
+    Returns
+    -------
+    DataArray
+        DataArray with shapely polygon per grid cell.
+    """
+    from shapely import Polygon
+
+    grid = ds.cf[["latitude", "longitude"]].load().reset_coords()
+    bounds = ds.cf.bounds
+
+    assert "latitude" in bounds
+    assert "longitude" in bounds
+    (lon_bounds,) = bounds["longitude"]
+    (lat_bounds,) = bounds["latitude"]
+
+    (points,) = xr.broadcast(grid)
+
+    bounds_dim = grid.cf.get_bounds_dim_name("latitude")
+    points = points.transpose(..., bounds_dim)
+    assert points.sizes[bounds_dim] == 2
+
+    lonbnd = points[lon_bounds].data
+    latbnd = points[lat_bounds].data
+
+    # geopandas needs this
+    expanded_lon = lonbnd[..., [0, 0, 1, 1]]
+    mask = expanded_lon[..., 0] >= 180
+    expanded_lon[mask, :] = expanded_lon[mask, :] - 360
+
+    # these magic numbers are OK :
+    # - 4 corners to a polygon, and
+    # - 2 from stacking lat, lon along the last axis
+    # flatten here to make iteration easier. It would be nice to avoid that.
+    # potentially with just np.vectorize. The polygon creation is the real slow bit.
+    # Shapely's MultiPolygon also iterates over a list in Python...
+    blocked = np.stack([expanded_lon, latbnd[..., [0, 1, 1, 0]]], axis=-1).reshape(
+        -1, 4, 2
+    )
+    polyarray = np.array(
+        [Polygon(blocked[i, ...]) for i in range(blocked.shape[0])], dtype="O"
+    )
```

Comment:
```suggestion
    polyarray = shapely.polygons(blocked)
```
This is how to avoid the loop. It gets about 10x faster.

Reply:
Requires shapely >= 2 but that should optimally be a general requirement now to avoid performance issues.

Reply:
I'm not sure I understand what comes before enough to suggest something complete, but for the curvilinear case I had this in another project:
```
ds = ds.cf.add_bounds(["longitude", "latitude"])
polyarray = shapely.polygons(shapely.linearrings(ds.lon_bounds, ds.lat_bounds))
```

Reply:
Amazing! This is why I love OSS =)

I looked at the docstrings for `Polygon`, `MultiPolygon` but didn't see a reference to `shapely.polygons`. It'd be nice if those functions were more prominent. While there is some user error here (I didn't see "Geometry Creation" under "API reference"), it's not clear at all why those functions are much faster.

Reply:
I'll say both suggestions seem to work. I'll go with @aulemahal's since it avoids the `np.stack`

Reply:
They are faster because you're shipping the whole array to C and let GEOS deal with the loop, while Polygon creates a scalar geometry. I'll bring the docs feedback to the upstream ;).